File: src/mvc/cpp/string_list_model/insert_element_dialog.cpp
#include "insert_element_dialog.h"
#include "./ui_insert_element_dialog.h"
#include <QPushButton>

InsertElementDialog::InsertElementDialog(QWidget *parent)
    : QDialog(parent), ui(new Ui::InsertElementDialog) {
  ui->setupUi(this);

  connect(this, &QDialog::accepted, this, [=]() {
    auto data = InsertElementData{ui->spinBox->value(), ui->lineEdit->text()};
    emit elementSelected(data);
    close();
  });

}

void InsertElementDialog::setMaxIndex(int maxIndex) {
  ui->spinBox->setMaximum(maxIndex);
}

InsertElementDialog::~InsertElementDialog() { delete ui; }

File: src/mvc/python/string_list_model/insert_element_dialog.py
from dataclasses import dataclass

from PyQt6 import uic, QtCore
from PyQt6.QtWidgets import QDialog


@dataclass
class InsertElementData:
    index: int
    text: str


class InsertElementDialog(QDialog):
    element_selected = QtCore.pyqtSignal(InsertElementData)

    def __init__(self, parent=None):
        super().__init__(parent)
        uic.loadUi("insert_element_dialog.ui", self)
        self.accepted.connect(self.on_accepted)

    def on_accepted(self):
        data = InsertElementData(self.spinBox.value(), self.lineEdit.text())
        self.element_selected.emit(data)
        self.close()

    def set_max_index(self, max_index):
        self.spinBox.setMaximum(max_index)
